core/ErrorQueue.h

```diff
@@ -17,7 +17,7 @@ class ErrorQueue {
     std::vector<std::unique_ptr<ErrorQueueMessage>> drainAll();
     std::vector<std::unique_ptr<ErrorQueueMessage>> drainFlushed();
     void collectForFile(core::FileRef whatFile, std::vector<std::unique_ptr<core::ErrorQueueMessage>> &out);
-    ErrorFlusher errorFlusher;
+    ErrorFlusherStdout errorFlusher;
```

Comment:
This should be an `ErrorFlusher` rather than the specific `ErrorFlusherStdout`. You can then instantiate it to an `ErrorFlusherStdout` instance in the `ErrorQueue` constructor. Later, we'll pass that responsibility to the callsite creating the error queue.

(You can make this a `unique_ptr` to make that work.)

Reply:
Why does it have to be `unique_ptr` to work?

Reply:
It has to be a pointer or reference type to work. The reason why has to do with determining the size of objects and storing things by value.

Most programming languages only support referring to objects by reference. For example, in Java, having a field like so:

```
class Blah {
ErrorFlusher errorFlusher;
}
```

...means that objects of type `Blah` have a field `errorFlusher` containing a _reference_ to an object with type `ErrorFlusher`. The size of that reference is constant -- it's the size of a pointer. Thus, we can store a reference to `ErrorFlusher` or any of its subclasses in the same amount of space.

In C++, this syntax has a different meaning. `ErrorFlusher errorFlusher` says that the object _contains_ a full `ErrorFlusher` object. In other words, the size of Blah in memory is roughly sizeof(ErrorFlusher) + sizeof(other fields on Blah).

Subclasses of `ErrorFlusher` may take more space than the base `ErrorFlusher` class, and, for that reason, cannot be stored by value as `ErrorFlusher`. Thus, we actually want to store a (smart) pointer to `ErrorFlusher` instead, which will let us store any subclass of `ErrorFlusher` in the field in the same amount of space.

There are other good reasons to use pointers over raw value types:

* **Reduces copying.** Assigning a new ErrorFlusher means that we are copying the contents of the first ErrorFlusher. Copying can be super cheap if ErrorFlusher only contains simple data types, but if it contains any object with a custom copy constructor then that code will have to run to copy that object.
  * tl;dr: We reserve referring to objects by value for super simple data types (like SymbolRef).
  * Fun note: Yeah, you can tell C++ how to copy your object.
* **Lets you forward declare to reduce #imports in headers.** Header files only need to know the full type information for types that are 1) stored by value in order to figure out the size of things in memory and 2) that are actually accessed in method implementations or field initialization in the header. References and pointer types are always the same size, so you can just tell the compiler with a forward decl that they exist and shave off #imports  (which can make compilation faster).